lib/cn/cn_tree.c

```diff
@@ -1370,26 +1370,27 @@ cn_node_comp_token_put(struct cn_tree_node *tn)
 static void
 cn_comp_release(struct cn_compaction_work *w)
 {
-    struct kvset_list_entry *le;
-    uint                     kx;
-
     assert(w->cw_node);
 
     if (w->cw_rspill_conc) {
+        struct cn_compaction_work *tmp HSE_MAYBE_UNUSED;
```

Comment:
This annotation doesn't seem right.

Reply:
Doh, it was right when it was an assert.  But I changed to abort because if this fails we really really don't want to proceeed...

Reply:
I agree with the abort() change definitely.

Reply:
Cool.  Fixed.  It's worse than I thought, if we hit the abort() then the kvdb is already corrupted.

Reply:
Doesn't every assert indicate an abort condition?

Reply:
Yes, but only in a debug build.  This is such a nasty condition that you want to halt if it happens.  It shouldn't happen if the concurrently rspill mechanisms are working, so it's just a fail safe in case there's some case we missed or someone breaks the code down the line...

Reply:
My point is all asserts are essentially halt conditions, and many can lead to corruption.  I can probably find dozens of asserts that lead to data corruption on release builds.  For example, an assert of an OMF struct magic value, or an assert of a loop control variable in a compaction merge loop.  Why not make them all aborts?

I'm playing devil's advocate here.  I prefer leaving asserts as asserts.  If you don't think it should be an assert, make it an error condition and handle it.

Reply:
IMO it's folly reading data from disk and not verifying that the data has integrity.  But that can be pretty expensive, so it's often just not done. Suprisingly it works pretty well almost 100% of the time.  I once had a bad disk that just returned the wrong data, no indication of error.  Took a while to track that down., I think it was a problem with the disk caching system rather than reading garbage from the platter, but we'll never know..

In the old days if you were editing a file on Solaris and the system crashed the file would be empty when the system came back up, which maybe is better than a corrupted file, but nonetheless irritating.  Had it been a critical file you'd have a big repair job on your hands.  The guys at Veritas would walk through fsck inode by inode after a crash to see how critical the correction would be (say, vs fsck -y).  I've never had the patience for that!!!

Anyway, I've changed it back to assert :-)